File: 41和为s的连续正数序列.java
// Brute Force O(n^2)
// import java.util.ArrayList;
// public class Solution {
//     public ArrayList<ArrayList<Integer> > FindContinuousSequence(int sum) {
//         ArrayList<ArrayList<Integer>> res = new ArrayList();
//         for(int i = 1; i < sum; i++) {
//             int temp = 0;
//             int j = i;
//             while(temp < sum) {
//                 temp += j;
//                 j++;
//             }
//             if(temp == sum) {
//                 ArrayList<Integer> list = new ArrayList();
//                 for(int k = i; k < j; k++) {
//                     list.add(k);
//                 }
//                 res.add(list);
//             }
//         }
//         return res;
//     }
// }

// two pointer methods or silde window method
// 思路见javascript解法 从js改写成java

import java.util.ArrayList;
public class Solution {
    public ArrayList<ArrayList<Integer> > FindContinuousSequence(int sum) {
        ArrayList<ArrayList<Integer>> res = new ArrayList();
        
        int low = 1, high = 2, current;
        while(low < high) {
            current = (high-low+1)*(high+low) / 2;
            if(current == sum) {
                ArrayList<Integer> temp = new ArrayList();
                for(int i = low; i <= high; i++) {
                    temp.add(i);
                }
                res.add(temp);
                low++;
            } else if(current < sum) {
                high++;
            } else {
                low++;
            }
        }
        return res;
    }
}

File: 剑指offer全解/41和为s的连续正数序列.js
// two pointer method based on arithmetic pregression sum
// 1   等差数列： current = (high-low+1)(high+low) / 2
// 2   初始化两个指针 low = 1, high = 2 覆盖尽可能多的情况
// 3   low < high 循环就可以继续进行
// 4.0 不断比较current 和 low
// 4.1 current == sum, 那么low~high之间的数满足要求，遍历其中的数存入一个数组 low++ 继续进行求下一个序列

function FindContinuousSequence(sum)
{
    // write code here
    var res = [];
    var low = 1; var high = 2;
    while(low < high) {
        current = (high-low+1)*(high+low) / 2;
        if(current == sum) {
            var list = [];
            for(let i = low; i <= high; i++) {
                list.push(i);
            }
            res.push(list);
            low++;
        } else if(current < sum) {
            high++;
        } else {
            low++;
        }
    }
    return res;
}